libioc/Provisioning/puppet.py

```diff
@@ -0,0 +1,233 @@
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""iocage provisioner for use with `puppet apply`."""
+import typing
+import os.path
+
+import git
+
+import libioc.errors
+import libioc.events
+import libioc.Pkg
+import libioc.Provisioning
+
+
+class ControlRepoUnavailableError(libioc.errors.IocException):
+    """Raised when the puppet control-repo is not available."""
+
+    def __init__(
+        self,
+        url: str,
+        reason: str,
+        logger: typing.Optional['libioc.Logger.Logger']=None
+    ) -> None:
+        msg = f"Puppet control-repo '{url}' is not available: {reason}"
+        libioc.errors.IocException.__init__(
+            self,
+            message=msg,
+            logger=logger
+        )
+
+
+class ControlRepoDefinition(dict):
+    """Puppet control-repo definition."""
+
+    _url: str
+    _name: str
+    _pkgs: typing.List[str]
+
+    def __init__(
+        self,
+        name: str,
+        url: str,
+        logger: 'libioc.Logger.Logger'
+    ) -> None:
+        self.logger = logger
+        self.url = url
+        self.name = name
+
+        _pkgs = ['puppet6']  # make this a Global Varialbe
+        if not (url.startswith('file://') or url.startswith('/')):
+            _pkgs += 'rubygem-r10k'
+
+    @property
+    def local(self) -> bool:
+        """Return whether this control repo resides locally."""
+        if not (self.url.startswith('file://') or self.url.startswith('/')):
+            return False
+        return True
+
+    @property
+    def remote(self) -> bool:
+        """Return whether this control repo resides locally."""
+        return not self.local
+
+    @property
+    def url(self) -> str:
+        """Return the Puppet Control-Repo URL."""
+        return str(self._url)
+
+    @url.setter
+    def url(self, value: str) -> None:
+        """Set the Puppet Control-Repo URL."""
+        self._url = value
+
+    @property
+    def name(self) -> str:
+        """Return the unique name for this Puppet Control-Repo URL."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str) -> None:
+        """Set a unique name for this Puppet Control-Repo URL."""
+        self._name = value
+
+    @property
+    def pkgs(self) -> typing.List[str]:
+        """Return list of packages required for this Provisioning method."""
+        return self._pkgs
+
+    def generate_postinstall(self) -> str:
+        """Return list of strings representing our postinstall."""
+        postinstall = """#!/bin/sh
+        set -eu
```

Comment:
we could also `set -x` here on appropriate log levels

Reply:
query logger level here and set these

Reply:
Just be verbose here. The logger will only print the command output on verbose levels.